README
Fprint version 0.00_01
======================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2015 by Iurii Shikin <shikin@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.16.3 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Fprint.

{{$NEXT}}

0.00_01  Wed Aug 19 13:41:21 2015
        - original version; created by h2xs 1.23 with options
            -A -B -b 5.8.9 -d -n Fprint --use-new-tests --skip-ppport --use-xsloader



Makefile.PL
use 5.008009;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Fprint',
    VERSION_FROM      => 'lib/Fprint.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Fprint.pm', # retrieve abstract from module
       AUTHOR         => 'Yuriy Shikin <yuriyshikin@gmail.com> <root@submin.ru>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
	# Insert -I. if you add *.h files later:
    INC               => '', # e.g., '-I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/Fprint.pm
package Fprint;

use 5.008009;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Fprint ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.00_01';
our $XS_VERSION = $VERSION;
$VERSION = eval $VERSION;  # see L<perlmodstyle>

require XSLoader;
XSLoader::load('Fprint', $XS_VERSION);

# Preloaded methods go here.

1;

__END__

# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Fprint - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Fprint;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Fprint, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Iurii Shikin <shikin@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Iurii Shikin <shikin@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.16.3 or,
at your option, any later version of Perl 5 you may have available.


=cut


daemon/dfprintd.c
/*
 * Dumb frintd daemon
 *
 * Copyright (C) 2015 Iurii Shikin <shikin@reg.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "dfprintd.h"

#include <yaml.h>

void sigterm(int signo)
{
    syslog(LOG_INFO, "daemon terminating");
    exit(0);
}

void sighup(int signo)
{
//    if(read_config())
        syslog(LOG_INFO, "config successfully reloaded");
}

static void
err_doit(int errnoflag, int error, const char *fmt, va_list ap)
{
    char buf[MAXLINE];

    vsnprintf(buf, MAXLINE, fmt, ap);
    if (errnoflag)
        snprintf(buf+strlen(buf), MAXLINE-strlen(buf), ": %s", strerror(error));
    strcat(buf, "\n");
    fflush(stdout); /* в случае, когда stdout и stderr – */
    /* одно и то же устройство */
    fputs(buf, stderr);
    fflush(NULL); /* сбрасывает все выходные потоки */
}

void
err_quit(const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    err_doit(0, 0, fmt, ap);
    va_end(ap);
    exit(1);
}

int
daemonize(void)
{
    int fd, fd0, fd1, fd2;
    pid_t pid;
    struct rlimit rl;
    struct sigaction sa;

    /* Сбросить маску режима создания файла */
    umask(0);

    /* Стать лидером новой сессии, чтобы утратить управляющий терминал */
    if ((pid = fork()) < 0)
        /* здесь отладочный вывод */
        return(-1);
    else if (pid != 0) /* родительский процесс */
        /* здесь отладочный вывод */
        return(0);

    setsid();

    /* Обеспечить невозможность обретения управляющего терминала в будущем */
    sa.sa_handler = SIG_IGN;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    if (sigaction(SIGHUP, &sa, NULL) < 0)
        /* здесь отладочный вывод */
        return(-1);

    if ((pid = fork()) < 0)
        /* здесь отладочный вывод */
        return(-1);
    else if (pid != 0) /* родительский процесс */
        /* здесь отладочный вывод */
        return(0);

    /* Закрыть все открытые файловые дескрипторы */
    if (getrlimit(RLIMIT_NOFILE, &rl) < 0)
        /* здесь отладочный вывод */
        return(-1);

    if (rl.rlim_max == RLIM_INFINITY)
        rl.rlim_max = 1024;

    for (fd = 0; fd < rl.rlim_max; fd++)
        close(fd);

    /* Присоединить файловые дескрипторы 0, 1 и 2 к /dev/null */
    fd0 = open("/dev/null", O_RDWR);
    fd1 = dup(0);
    fd2 = dup(0);

    /* Инициализировать файл журнала */
    if (fd0 != 0 || fd1 != 1 || fd2 != 2) {
        /* здесь отладочный вывод */
        return(-1);
    }

    /*
     * Назначить корневой каталог текущим рабочим каталогом,
     * чтобы впоследствии можно было отмонтировать файловую систему.
     */
    if (chdir("/") < 0)
        /* здесь отладочный вывод */
        return(-1);

    syslog(LOG_INFO, "daemon successfully started"); // напишем системный лог
    return(1);
}

int
already_running(void)
{
    int fd;
    struct flock lock;

    lock.l_type = F_WRLCK;     /* F_RDLCK, F_WRLCK, F_UNLCK */
    lock.l_start = 0;  /* смещение в байтах относительно l_whence */
    lock.l_whence = SEEK_SET; /* SEEK_SET, SEEK_CUR, SEEK_END */
    lock.l_len = 1;       /* количество байт (0 – до конца файла) */

    fd = open(LOCKFILE, O_RDWR|O_CREAT, LOCKMODE);
    if (fd < 0) {
        syslog(LOG_ERR, "Can't open lockfile %s: %s", LOCKFILE, strerror(errno));
        return(-1);
    }
    if (fcntl((fd), F_SETLK, &lock) < 0) {
        if (errno == EACCES || errno == EAGAIN) {
            close(fd);
            return(1);
        }
        syslog(LOG_ERR, "Can't locking %s: %s", LOCKFILE, strerror(errno));
        return(-1);
    }
//    ftruncate(fd, 0);
    return(0);
}

void
read_config(struct config * conf, char * file)
{
    int state = 0;
    char *tk;

    int ** ppdatai;
    char ** ppdatac;

    yaml_parser_t parser;
    yaml_token_t  token;

    conf->config = file;

    FILE *fh = fopen(conf->config, "r");

    /* Initialize parser */
    if(!yaml_parser_initialize(&parser))
        fputs("Failed to initialize parser!\n", stderr);
    if(fh == NULL)
        fputs("Failed to open file!\n", stderr);

    /* Set input file */
    yaml_parser_set_input_file(&parser, fh);

    /* BEGIN new code */
    do {
        yaml_parser_scan(&parser, &token);
        switch(token.type)
        {
            /* Token types (read before actual token) */
            case YAML_KEY_TOKEN:   state = 0; break;
            case YAML_VALUE_TOKEN: state = 1; break;
            case YAML_SCALAR_TOKEN:
                tk = token.data.scalar.value;
                if (state == 0) {
                    if (!strcmp(tk, "socket")) {
                        ppdatac = &conf->socket;
                    } else if (!strcmp(tk, "pidfile")) {
                        ppdatac = &conf->pidfile;
                    } else if (!strcmp(tk, "debug")) {
                        ppdatai = &conf->debug;
                    } else if (!strcmp(tk, "workers")) {
                        ppdatai = &conf->workers;
                    } else {
                        printf("Unrecognised key: %s\n", tk);
                    }
                } else {
                    *ppdatac = strdup(tk);
                }
                break;
            default: break;
        }
        if(token.type != YAML_STREAM_END_TOKEN)
            yaml_token_delete(&token);
    } while(token.type != YAML_STREAM_END_TOKEN);
    yaml_token_delete(&token);
    /* END new code */

    /* Cleanup */
    yaml_parser_delete(&parser);
    fclose(fh);
}

void
show_usage(void)
{
    printf("Usage:\n %s [oprtions]", NAME);
    printf("\n");
    printf("\nOptions:");
    printf("\n -c|--config <path>  full path to file configuration");
    printf("\n -d|--debug          start daemon with debug mode enable");
    printf("\n -v|--version        show version and exit");
    printf("\n -h|--help           show this help and exit");
    printf("\n");
}

void
show_version(void)
{
    printf("%s %s", NAME, VERSION);
    printf("\n");
}






daemon/dfprintd.h
/*
 * dfprintd header file
 * Copyright (C) 2015 Iurii Shikin <shikin@reg.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef _DFPRINTD_H
#define _DFPRINTD_H

#include <stdarg.h>
#include <sys/types.h>
#include <stdlib.h>
#include <syslog.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/resource.h>
#include <signal.h>
#include <errno.h>
#include <sys/stat.h>
#include <fcntl.h>

/* general */
#define NAME "dfprintd"
#define VERSION "0.0.1"

int daemonize(void);
//int already_running(void);

/* config */
#define CONFFILE "/etc/dfprintd.conf"
#define DEBUG 0
#define MAXPATHL 255
#define WORKERS 5
#define SOCKET "/var/run/dfprintd.socket"
#define PIDFILE "/var/run/dfprintd.pid"
#define MAXFILEL 255

struct config {
    char * socket;
    char * pidfile;
    char * config;
    int * workers;
    int * debug;
};
void read_config(struct config *, char *);

/* locks */
#define LOCKFILE "/var/run/dfprintd.lock"
#define LOCKMODE (S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH)

/* signals */
void sigterm(int);
void sighup(int);

/* errors */
#define MAXLINE 4096
static void err_doit(int, int, const char *, va_list);
void err_quit(const char *, ...);

/* other */
void show_usage(void);
void show_version(void);

#endif /* dfprintd.h */


daemon/main.c
/*
 * dfprintd Dumb frintd daemon
 *
 * Copyright (C) 2015 Iurii Shikin <shikin@reg.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "dfprintd.h"

#include <getopt.h>
#include <talloc.h>

int
main(int argc, char **argv)
{
    struct sigaction sa;
    int opt, daemon;
    int debug, workers;

    struct config *conf = talloc(NULL, struct config);

    /* Init default configuration */
    conf->socket = SOCKET;
    conf->pidfile = PIDFILE;
    conf->config = CONFFILE;

    debug = DEBUG;
    workers = WORKERS;
    conf->debug = &debug;
    conf->workers = &workers;
printf("\nAfter load conf:\n\tsocket: %s\n\tpidfile: %s\n\tconfig: %s\n\tworkers: %d\n\tdebug: %d\n",
    conf->socket,
    conf->pidfile,
    conf->config,
    *conf->workers,
    *conf->debug
);
    while(1){
        int opt_index;
        static struct option long_opt[] = {
            {"config", 1, 0, 'c'},
            {"debug", 0, 0, 'd'},
            {"help", 0, 0, 'h'},
            {"version", 0, 0, 'v'},
            {0,0,0,0}
        };

        if((opt = getopt_long(argc, argv, "c:dhv", long_opt, &opt_index)) == -1)
            break;

        switch(opt){
            case 'c':
                read_config(conf, optarg);
                break;
            case 'd':
                exit(0);
            case 'h':
                show_usage();
                exit(0);
            case 'v':
                show_version();
                exit(0);
            default:
                show_usage();
                exit(-1);
        }
    }

    /* Убедиться в том, что ранее не была запущена другая копия демона 
    if (already_running()) {
        syslog(LOG_ERR, "daemon already is running");
        exit(1);
    }
*/
    /* Перейти в режим демона 
    daemon = daemonize();
    if (daemon < 0)
        exit(-1);
    else if (daemon == 0)
        exit(0);
*/

    /* Установить обработчики сигналов */
    sa.sa_handler = sigterm;
    sigemptyset(&sa.sa_mask);
    sigaddset(&sa.sa_mask, SIGHUP);
    sa.sa_flags = 0;
    if (sigaction(SIGTERM, &sa, NULL) < 0) {
        syslog(LOG_ERR, "невозможно перехватить сигнал SIGTERM: %s", strerror(errno));
        exit(1);
    }

    sa.sa_handler = sighup;
    sigemptyset(&sa.sa_mask);
    sigaddset(&sa.sa_mask, SIGTERM);
    sa.sa_flags = 0;
    if (sigaction(SIGHUP, &sa, NULL) < 0) {
        syslog(LOG_ERR, "невозможно перехватить сигнал SIGHUP: %s", strerror(errno));
        exit(1);
    }

//    for(;;);

    exit(0);
}


/*
printf("\nAfter load conf:\n\tsocket: %s\n\tpidfile: %s\n\tconfig: %s\n\tworkers: %d\n\tdebug: %d\n",
    conf->socket,
    conf->pidfile,
    conf->config,
    conf->workers,
    conf->debug
);

*/